File: backend/app/schemas/practice_loop_stats.py
"""Practice Loop Stats schemas — #512.

Spec: docs/specs/practice/youtube_loop_practice_spec.md §4 (선생님 통계) + §5 (동기화).

Three flows:
- Student sync (POST): client uploads one or more (section, count, last_played)
  rows. Idempotent — server keeps max(stored, incoming).
- Teacher list (GET): query a single student's per-section repeats with an
  optional time window (`weekly` / `monthly`).
- Teacher list (GET aggregate per teacher): all students summarised — used
  for the dashboard entry card on the home screen.
"""

from __future__ import annotations

import datetime as _dt
import enum

from pydantic import BaseModel, ConfigDict, Field


class StatsWindow(str, enum.Enum):
    """Time window toggle exposed in the teacher UI (spec §4.3)."""

    weekly = "weekly"
    monthly = "monthly"


# ---------------------------------------------------------------------------
# Student sync — POST /api/v1/practice-loop-stats/sync
# ---------------------------------------------------------------------------


class PracticeLoopStatsSyncEntry(BaseModel):
    """One (section, count) pair the client wants to persist."""

    section_id: str = Field(..., min_length=1, max_length=64)
    repeat_count: int = Field(..., ge=0)
    last_played_at: _dt.datetime


class PracticeLoopStatsSyncRequest(BaseModel):
    """Batched upload from the client. Sent at session end / scheduled flush."""

    entries: list[PracticeLoopStatsSyncEntry] = Field(default_factory=list)


class PracticeLoopStatsSyncResult(BaseModel):
    """Sync outcome — server reports how many rows were upserted vs skipped."""

    upserted: int = 0
    skipped: int = 0
    rejected: int = 0


# ---------------------------------------------------------------------------
# Teacher list — GET /api/v1/practice-loop-stats/students/{student_id}
# ---------------------------------------------------------------------------


class PracticeLoopStatsRow(BaseModel):
    """One section row for the teacher view."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    student_id: str
    teacher_id: str
    section_id: str
    repeat_count: int
    last_played_at: _dt.datetime
    created_at: _dt.datetime
    updated_at: _dt.datetime | None = None
    # Convenience labels (populated by the service when joinable).
    piece_name: str | None = None
    section_name: str | None = None


class PracticeLoopStatsListResponse(BaseModel):
    """Per-student response for the teacher UI."""

    student_id: str
    window: StatsWindow
    total_repeats: int = 0
    rows: list[PracticeLoopStatsRow] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Teacher dashboard summary — GET /api/v1/practice-loop-stats/summary
# ---------------------------------------------------------------------------


class PracticeLoopStatsStudentSummary(BaseModel):
    """One student's roll-up for the teacher dashboard card."""

    student_id: str
    student_name: str | None = None
    total_repeats: int = 0
    last_played_at: _dt.datetime | None = None


class PracticeLoopStatsSummaryResponse(BaseModel):
    """Per-teacher roll-up across all linked students."""

    teacher_id: str
    window: StatsWindow
    students: list[PracticeLoopStatsStudentSummary] = Field(default_factory=list)

File: frontend/lib/features/practice/domain/entities/practice_loop_stats.dart
/// Aggregated per-section repeat counts for the teacher dashboard (#512).
///
/// Spec: docs/specs/practice/youtube_loop_practice_spec.md §4 (선생님 통계).
///
/// One row per (student, section) pair. ``repeatCount`` is the cumulative
/// total the student has played for that section across all sessions.
class PracticeLoopStats {
  /// Backend row id.
  final String id;

  /// Student profile id (NOT the user id — these come from the server).
  final String studentId;

  /// Teacher profile id (owner of the student).
  final String teacherId;

  /// Practice section id this row aggregates.
  final String sectionId;

  /// Cumulative repeat count reported by the client.
  final int repeatCount;

  /// Latest known play time across all syncs.
  final DateTime lastPlayedAt;

  /// Piece label hydrated by the server when available (UI sugar).
  final String? pieceName;

  /// Section label hydrated by the server when available (UI sugar).
  final String? sectionName;

  const PracticeLoopStats({
    required this.id,
    required this.studentId,
    required this.teacherId,
    required this.sectionId,
    required this.repeatCount,
    required this.lastPlayedAt,
    this.pieceName,
    this.sectionName,
  });
}

/// Roll-up of one student's repeats for the dashboard summary card.
class StudentRepeatStats {
  final String studentId;
  final String? studentName;
  final int totalRepeats;
  final DateTime? lastPlayedAt;

  const StudentRepeatStats({
    required this.studentId,
    required this.totalRepeats,
    this.studentName,
    this.lastPlayedAt,
  });
}

/// Time-window toggle for the teacher view. Spec §4.3.
enum PracticeLoopStatsWindow {
  weekly,
  monthly;

  /// Wire value sent to the backend (`?window=...`).
  String get wireName => switch (this) {
    PracticeLoopStatsWindow.weekly => 'weekly',
    PracticeLoopStatsWindow.monthly => 'monthly',
  };
}

/// Server-side outcome of a sync request — exposed for instrumentation/tests.
class PracticeLoopStatsSyncResult {
  final int upserted;
  final int skipped;
  final int rejected;

  const PracticeLoopStatsSyncResult({
    this.upserted = 0,
    this.skipped = 0,
    this.rejected = 0,
  });

  /// Convenience flag for the offline-queue retry path.
  bool get hadAnyServerWrites => upserted > 0;
}

/// Pending sync entry persisted in the offline Hive queue.
///
/// The queue is flushed at session end or when connectivity returns; the
/// server treats the upload as idempotent so re-flush is safe.
class PendingLoopStatsSync {
  final String sectionId;
  final int repeatCount;
  final DateTime lastPlayedAt;

  const PendingLoopStatsSync({
    required this.sectionId,
    required this.repeatCount,
    required this.lastPlayedAt,
  });

  Map<String, dynamic> toJson() => {
    'sectionId': sectionId,
    'repeatCount': repeatCount,
    'lastPlayedAt': lastPlayedAt.toIso8601String(),
  };

  static PendingLoopStatsSync fromJson(Map<String, dynamic> json) {
    return PendingLoopStatsSync(
      sectionId: json['sectionId'] as String,
      repeatCount: json['repeatCount'] as int,
      lastPlayedAt: DateTime.parse(json['lastPlayedAt'] as String),
    );
  }
}
